indexer/app.py

```diff
@@ -107,28 +116,94 @@ def load_config_class():
 
+dss_url = os.environ['AZUL_DSS_ENDPOINT']
 loaded_properties = indexer_properties(dss_url, **kwargs)
 loaded_indexer = indexer_to_load(loaded_properties)
 
+num_workers = int(os.environ['AZUL_INDEX_WORKERS'])
+num_dss_workers = int(os.environ['AZUL_DSS_WORKERS'])
+requests.adapters.DEFAULT_POOLSIZE = num_workers * num_dss_workers
+
 
 @app.route('/', methods=['POST'])
 def post_notification():
     """
-    Receive the notification event.
-
-    This function takes in a POST event from the blue box and
-    triggers the whole indexing process.
-
+    Receive a notification event and either queue it for asynchronous indexing or process it synchronously.
     """
-    # blue box sends json with post request; get payload
-    payload = app.current_request.json_body
-    app.log.info("Received notification %s", payload)
-    # Index based on the payload
-    loaded_indexer.index(payload)
+    notification = app.current_request.json_body
+    log.info("Received notification %r", notification)
+    params = app.current_request.query_params
+    if params and params.get('sync', 'False').lower() == 'true':
+        loaded_indexer.index(notification)
+    else:
+        queue().send_message(MessageBody=json.dumps(notification))
+        log.info("Queued notification %r", notification)
     return {"status": "done"}
 
 
 @app.route('/escheck')
 def es_check():
     """Check the status of ElasticSearch by returning its info."""
     return json.dumps(loaded_properties.elastic_search_client.info())
+
+
+# Work around https://github.com/aws/chalice/issues/856
+
+def new_handler(self, event, context):
+    app.lambda_context = context
+    return old_handler(self, event, context)
+
+
+old_handler = chalice.app.ScheduledEventHandler.__call__
+chalice.app.ScheduledEventHandler.__call__ = new_handler
+
+
+@app.schedule("rate(1 minute)", name='worker')
+def index(event):
+    log.info(f'Starting worker threads')
+    remaining_time = RemainingLambdaContextTime(app.lambda_context)
+    with ThreadPoolExecutor(num_workers) as tpe:
+        futures = [tpe.submit(_index, i, remaining_time) for i in range(num_workers)]
+        for future in as_completed(futures):
+            e = future.exception()
+            if e:
+                log.error("Exception in worker thread", exc_info=e)
+    log.info(f'Shutting down')
+
+
+def queue():
+    session = boto3.session.Session()  # See https://github.com/boto/boto3/issues/801
+    queue_name = "azul-notify-" + os.environ['AZUL_DEPLOYMENT_STAGE']
+    queue = session.resource("sqs").get_queue_by_name(QueueName=queue_name)
+    return queue
+
+
+def _index(worker, remaining_time) -> None:
+    _queue = queue()
+    polling_time = 20
+    indexing_time = 60
+    shutdown_time = 5
+    backoff_time = 10
+    while polling_time + shutdown_time + indexing_time < remaining_time.get():
+        visibility_timeout = remaining_time.get() + backoff_time
+        messages = _queue.receive_messages(MaxNumberOfMessages=1,
```

Comment:
It is technically a list, albeit a singleton one. It would be counter-intuitive to have a variable be a list of things but name it using the singular. Once I extract the single element into a new variable a few lines down, that variable *is* called `message`.